libdrgn/iterators.c

```diff
@@ -0,0 +1,461 @@
+	bool node_is_not_null = false;
+	struct drgn_error *err = NULL;
+	do {
+		err = iter->started ?
+				    radix_tree_iter_run_frame(iter, &is_exhausted) :
+				    radix_tree_iter_handle_node(iter, &iter->root, 0);
+		iter->started = true;
+		if (err)
+			return err;
+		err = drgn_object_bool(&iter->entry.node, &node_is_not_null);
+	} while (!is_exhausted && !err && !node_is_not_null);
+	*ret = is_exhausted || err ? NULL : &iter->entry;
+	return err;
+}
+
+void radix_tree_iter_deinit(struct radix_tree_iter *iter)
+{
+	drgn_object_deinit(&iter->root);
+	drgn_object_deinit(&iter->entry.node);
+	while (iter->frames.size) {
+		drgn_object_deinit(&frames_pop(&iter->frames)->slots);
+	}
+	frames_deinit(&iter->frames);
+}
+
+struct drgn_error *idr_iter_init(const struct drgn_object *idr,
+				 struct idr_iter *ret)
+{
+	struct drgn_object idr_rt;
+	struct drgn_object idr_base;
+	struct drgn_error *err;
+	drgn_object_init(&idr_rt, drgn_object_program(idr));
+	drgn_object_init(&idr_base, drgn_object_program(idr));
+
+	err = drgn_object_member(&idr_base, idr, "idr_base");
+	if (!err) {
+		err = drgn_object_read_unsigned(&idr_base, &ret->base);
+		if (err)
+			goto out;
+	} else if (err->code == DRGN_ERROR_LOOKUP) {
+		drgn_error_destroy(err);
+		err = NULL;
+		ret->base = 0;
+	} else {
+		goto out;
+	}
+
+	err = drgn_object_member(&idr_rt, idr, "idr_rt");
+	if (err)
+		goto out;
+
+	err = radix_tree_iter_init(drgn_object_program(idr), &idr_rt,
+				   &ret->iter);
+out:
+	drgn_object_deinit(&idr_rt);
+	drgn_object_deinit(&idr_base);
+	return err;
+}
+
+struct drgn_error *idr_iter_next(struct idr_iter *iter, idr_iter_entry_t **ret)
+{
+	struct drgn_error *err = radix_tree_iter_next(&iter->iter, ret);
+	if (!err && *ret)
+		(*ret)->index += iter->base;
+	return err;
+}
+
+void idr_iter_deinit(struct idr_iter *iter)
+{
+	radix_tree_iter_deinit(&iter->iter);
+}
+
+static struct drgn_error *pid_iter_init_old_kernel(struct drgn_program *prog,
+						   const struct drgn_object *ns,
+						   struct pid_iter *ret)
+{
+	struct drgn_error *err;
+	ret->has_idr = false;
+	ret->index = -1;
+	drgn_object_init(&ret->pid_hash, prog);
+	drgn_object_init(&ret->pos, prog);
+	drgn_object_init(&ret->ns, prog);
+	drgn_object_init(&ret->entry, prog);
+	err = drgn_program_find_object(prog, "pid_hash", NULL,
+				       DRGN_FIND_OBJECT_ANY, &ret->pid_hash);
+	if (err)
+		return err;
+	struct drgn_qualified_type void_star_type;
+	err = drgn_program_find_type(prog, "void *", NULL, &void_star_type);
+	if (err)
+		return err;
+	err = drgn_object_set_unsigned(&ret->pos, void_star_type, 0, 0);
+	if (err)
+		return err;
+	err = drgn_object_copy(&ret->ns, ns);
+	if (err)
+		return err;
+	err = drgn_program_find_type(prog, "struct pid", NULL, &ret->pid_type);
+	if (err)
+		return err;
+	err = drgn_program_find_type(prog, "struct upid", NULL,
+				     &ret->upid_type);
+	if (err)
+		return err;
+	struct drgn_object pidhash_shift;
+	drgn_object_init(&pidhash_shift, prog);
```

Comment:
This also needs to be deinitialized.

Reply:
Sorry, I was under the impression that `drgn_objects` which represent primitive data like integers didn't need to be deinitialized. Is the logic behind deinitializing them the same as for your suggestion to deinitialize contants?

Reply:
There are a few reasons:

1. Like I mentioned for constants, although drgn doesn't currently allocate anything that needs to be deinitialized for integers, that could change.
2. The variable isn't guaranteed to be an integer: a newer kernel version, or corrupted debug info, could make it something that does need to be freed.
3. At some point, I'd like to implement an "object sanitizer" to check for object leaks. It's easier to just enforce that for all objects regardless of type.